dev-tools/mage/build.go

```diff
@@ -0,0 +1,144 @@
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package mage
+
+import (
+	"fmt"
+	"go/build"
+	"log"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/magefile/mage/sh"
+	"github.com/pkg/errors"
+)
+
+// BuildArgs are the arguments used for the "build" target and they define how
+// "go build" is invoked.
+type BuildArgs struct {
+	Name       string // Name of binary. (On Windows '.exe' is appended.)
+	OutputDir  string
+	CGO        bool
+	Static     bool
+	Env        map[string]string
+	LDFlags    []string
+	Vars       map[string]string // Vars that are passed as -X key=value with the ldflags.
+	ExtraFlags []string
+}
+
+// DefaultBuildArgs returns the default BuildArgs for use in builds.
+func DefaultBuildArgs() BuildArgs {
+	args := BuildArgs{
+		Name: BeatName,
+		CGO:  build.Default.CgoEnabled,
+		Vars: map[string]string{
+			"github.com/elastic/beats/libbeat/version.buildTime": "{{ date }}",
+			"github.com/elastic/beats/libbeat/version.commit":    "{{ commit }}",
+		},
+	}
+
+	repo, err := GetProjectRepoInfo()
+	if err != nil {
+		panic(errors.Wrap(err, "failed to determine project repo info"))
+	}
+
+	if !repo.IsElasticBeats() {
+		// Assume libbeat is vendored and prefix the variables.
+		prefix := repo.RootImportPath + "/vendor/"
+		prefixedVars := map[string]string{}
+		for k, v := range args.Vars {
+			prefixedVars[prefix+k] = v
+		}
+		args.Vars = prefixedVars
+	}
+
+	return args
+}
+
+// DefaultGolangCrossBuildArgs returns the default BuildArgs for use in
+// cross-builds.
+func DefaultGolangCrossBuildArgs() BuildArgs {
+	args := DefaultBuildArgs()
+	args.Name += "-" + Platform.GOOS + "-" + Platform.Arch
+	args.OutputDir = filepath.Join("build", "golang-crossbuild")
+	if bp, found := BuildPlatforms.Get(Platform.Name); found {
+		args.CGO = bp.Flags.SupportsCGO()
+	}
+	return args
+}
+
+// GolangCrossBuild invokes "go build" inside of the golang-crossbuild Docker
+// environment.
+func GolangCrossBuild(params BuildArgs) error {
+	if os.Getenv("GOLANG_CROSSBUILD") != "1" {
+		return errors.New("Use the crossBuild target. golangCrossBuild can " +
+			"only be executed within the golang-crossbuild docker environment.")
+	}
+
+	defer dockerChown(filepath.Join(params.OutputDir, params.Name+binaryExtension(GOOS)))
+	return Build(params)
+}
+
+// Build invokes "go build" to produce a binary.
+func Build(params BuildArgs) error {
+	fmt.Println(">> build: Building", params.Name)
+
+	binaryName := params.Name + binaryExtension(GOOS)
+
+	if params.OutputDir != "" {
+		if err := os.MkdirAll(params.OutputDir, 0755); err != nil {
+			return err
+		}
+	}
+
+	// Environment
+	env := params.Env
+	if env == nil {
+		env = map[string]string{}
+	}
+	cgoEnabled := "0"
```

Comment:
If I remember correctly we use now CGO by default. Should we also set it here to 1 by default? At the same time I like that community beats can easily built without it.

Reply:
The passed in `BuildArgs` created by `DefaultBuildArgs()` and `DefaultGolangCrossBuildArgs` will set the cgo flag appropriately. (it's a bit complex)

- `mage build` (equivalent of `make` or `go build`) - This uses `DefaultBuildArgs()` as its starting point (Beats can further override the args). The logic for enabling cgo is exactly the same as it is for Go because it reuses the stdlib logic. Basically the logic is `if env[GOOS] == runtime.GOOS` then CGO_ENABLED=1. This ensures that cross-builds (`GOOS=netbsd mage build`) have CGO turned off since cross-compilers are not typically installed.

- `mage crossBuild` - The defaults come from `DefaultGolangCrossBuildArgs()`. This is the target that builds the binaries inside of docker using [golang-crossbuild](https://github.com/elastic/golang-crossbuild). CGO is enabled here by default (generally). More precisely if the target platform has CGO support then it will be enabled (that list came from `go tool dist list -json`).